sapl/sessao/views.py

```diff
@@ -4384,6 +4384,90 @@ class DeleteView(MasterDetailCrud.DeleteView):
         pass
 
 
+class LeituraEmBlocoExpediente(PermissionRequiredForAppCrudMixin, ListView):
+    template_name = 'sessao/leitura/leitura_bloco.html'
+    app_label = AppConfig.label
+    expediente = True
+    paginate_by = 100
+
+    def get_queryset(self):
+        return ExpedienteMateria.objects.filter(sessao_plenaria_id=self.kwargs['pk'],
+                                                retiradapauta=None, tipo_votacao=4, registroleitura__materia=None)
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context['pk'] = self.kwargs['pk']
+        context['root_pk'] = self.kwargs['pk']
+        if not verifica_sessao_iniciada(self.request, self.kwargs['pk']):
+            context['sessao_iniciada'] = False
+            return context
+        context['sessao_iniciada'] = True
+        context['turno_choices'] = Tramitacao.TURNO_CHOICES
+        context['title'] = SessaoPlenaria.objects.get(id=self.kwargs['pk'])
+        if self.expediente:
+            context['expediente'] = True
+        else:
+            context['expediente'] = False
+        return context
+
+    def post(self, request, *args, **kwargs):
+        if 'marcadas_4' in request.POST:
+            models = None
+            if request.POST['origem'] == 'ordem':
+                models = OrdemDia.objects.filter(id__in=request.POST.getlist('marcadas_4'))
+            elif request.POST['origem'] == 'expediente':
+                models = ExpedienteMateria.objects.filter(id__in=request.POST.getlist('marcadas_4'))
+
+            if models is None:
+                messages.add_message(self.request, messages.ERROR,
+                                     _('Impossível localizar as matérias selecionadas'))
+                return self.get(request, self.kwargs)
+            else:
+                for m in models:
```

Comment:
Muita calma nessa hora... :) Eles tão pedindo isso pra poder fazer leitura de 400 matérias ou mais! O **ideal** é tentar potencializar a quantidade de operações em lote o máximo que pudermos pra fazer o menor número possível de chamadas a BD. Neste sentido, eu sugiro que exploremos o uso de bulk_update, bulk_delete e bulk_create (um ou mais desses não está disponível no Django 2, se bem me lembro 😢). Considerando isso a gente pode fazer uma lógica mais ou menos assim:

DELETAR EM LOTE
```
materias = [m.materia for m in models]

RegistroLeitura.objects.filter(materia__in=materias, ...).delete()
```

Fazer o FOR pra criar os objetos novos

```
leituras = []
for m in models:
     obj = RegistroLeitura(ordem=m, materia=materia,
                           observacao=request.POST['observacao'],
                           user=self.request.user,
                           ip=get_client_ip(self.request)) 
                           // ou expediente=m Ah! E pode colocar tudo dentro do objeto, não precisa ser um por linha
     leituras.append(obj)
```

FORA do loop fazer um bulk_create: `RegistroLeituraz.objects.bulk_create(leituras)`